crates/ruma-federation-api/src/authenticated_media.rs

```diff
@@ -0,0 +1,519 @@
+
+    let mut body_writer = T::default().writer();
+
+    // Add first boundary separator and header for the metadata.
+    let _ = write!(
+        body_writer,
+        "\r\n--{boundary}\r\n{}: {}\r\n\r\n",
+        http::header::CONTENT_TYPE,
+        mime::APPLICATION_JSON
+    );
+
+    // Add serialized metadata.
+    serde_json::to_writer(&mut body_writer, metadata)?;
+
+    // Add second boundary separator.
+    let _ = write!(body_writer, "\r\n--{boundary}\r\n");
+
+    // Add content.
+    match content {
+        FileOrLocation::File(content) => {
+            // Add headers.
+            let content_type =
+                content.content_type.as_deref().unwrap_or(mime::APPLICATION_OCTET_STREAM.as_ref());
+            let _ = write!(body_writer, "{}: {content_type}\r\n", http::header::CONTENT_TYPE);
+
+            if let Some(content_disposition) = &content.content_disposition {
+                let _ = write!(
+                    body_writer,
+                    "{}: {content_disposition}\r\n",
+                    http::header::CONTENT_DISPOSITION
+                );
+            }
+
+            // Add empty line separator after headers.
+            let _ = body_writer.write_all(b"\r\n");
+
+            // Add bytes.
+            let _ = body_writer.write_all(&content.file);
+        }
+        FileOrLocation::Location(location) => {
+            // Only add location header and empty line separator.
+            let _ = write!(body_writer, "{}: {location}\r\n\r\n", http::header::LOCATION);
+        }
+    }
+
+    // Add final boundary.
+    let _ = write!(body_writer, "\r\n--{boundary}--");
+
+    let content_type = format!("{MULTIPART_MIXED}; boundary={boundary}");
+    let body = body_writer.into_inner();
+
+    Ok(http::Response::builder().header(http::header::CONTENT_TYPE, content_type).body(body)?)
+}
+
+/// Deserialize the given metadata and content from a `http::Response` with a `multipart/mixed`
+/// body.
+#[cfg(feature = "client")]
+fn try_from_multipart_mixed_response<T: AsRef<[u8]>>(
+    http_response: http::Response<T>,
+) -> Result<
+    (ContentMetadata, FileOrLocation),
+    ruma_common::api::error::FromHttpResponseError<ruma_common::api::error::MatrixError>,
+> {
+    use ruma_common::api::error::{HeaderDeserializationError, MultipartMixedDeserializationError};
+
+    // First, get the boundary from the content type header.
+    let body_content_type = http_response
+        .headers()
+        .get(http::header::CONTENT_TYPE)
+        .ok_or_else(|| HeaderDeserializationError::MissingHeader("Content-Type".to_owned()))?
+        .to_str()?
+        .parse::<mime::Mime>()
+        .map_err(|e| HeaderDeserializationError::InvalidHeader(e.into()))?;
+
+    if !body_content_type.essence_str().eq_ignore_ascii_case(MULTIPART_MIXED) {
+        return Err(HeaderDeserializationError::InvalidHeaderValue {
+            header: "Content-Type".to_owned(),
+            expected: "multipart/mixed".to_owned(),
```

Comment:
Shouldn't we just use `MULTIPART_MIXED` here?

```suggestion
            expected: MULTIPART_MIXED.to_owned(),
```

Reply:
Applied